### PR TITLE
feat: add conduit endpoints

### PR DESCRIPTION
**Pull request by serycodes**
Type: Feature

Fixes: #557

### CODE REVIEW

File: packages/api/src/endpoints/eventSub/HelixEventSubConduitShard.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+	/** @private */
+	get _transport(): HelixEventSubTransportData {
+		return this[rawDataSymbol].transport;
+	}
+
+	/** @private */
+	set _status(status: HelixEventSubSubscriptionStatus) {
+		this[rawDataSymbol].status = status;
+	}
```
Comment (d-fischer) [inline packages/api/src/endpoints/eventSub/HelixEventSubConduitShard.ts:44]: What are these for?

File: packages/api/src/interfaces/endpoints/eventSub.external.ts [hunk 2]
```diff
@@ -33,18 +41,67 @@ export interface HelixEventSubWebSocketTransportData {
...
+	 * The transport method.
+	 */
+	transport: HelixEventSubTransportData;
+}
+
+/** @private */
+export type HelixEventSubTransportData =
+	| HelixEventSubWebHookTransportData
+	| HelixEventSubWebSocketTransportData
+	| HelixEventSubConduitTransportData;
```
Comment (d-fischer) [inline packages/api/src/interfaces/endpoints/eventSub.external.ts:104]: I don't think conduits can be nested into other conduits. If that's true, this should be reflected by this type (and the type name should also reflect its use here).

**Review by d-fischer — CHANGES_REQUESTED** [review]

File: packages/api/src/interfaces/endpoints/eventSub.external.ts (hunk 2 above)
Comment (serycodes) [reply packages/api/src/interfaces/endpoints/eventSub.external.ts:104]: Sorry, I'm not following here.
If I understand you correctly: Conduits are a third transport type now, along with webhook and websockets when calling /helix/eventsub/subscriptions. They made sense to add here.

Raw (and anonymized) response from Helix:

```
{
    "total": 1234,
    "data": [
    {
        "id": "random-guid",
        "status": "enabled",
        "type": "channel.chat.message",
        "version": "1",
        "condition": {
            "broadcaster_user_id": "1234",
            "user_id": "5678"
        },
        "created_at": "2024-03-08T02:29:13.67322332Z",
        "transport": {
            "method": "conduit",
            "conduit_id": "random-guid"
        },
        "cost": 0
    },
    {
        "id": "random-guid",
        "status": "enabled",
        "type": "channel.hype_train.progress",
        "version": "1",
        "condition": {
            "broadcaster_user_id": "1234"
        },
        "created_at": "2021-10-18T19:51:04.304359423Z",
        "transport": {
            "method": "webhook",
            "callback": "https://my.callback/url"
        },
        "cost": 0
    }
],
    "max_total_cost": 10000,
    "total_cost": 1234,
    "pagination": {}
}
```

File: packages/api/src/endpoints/eventSub/HelixEventSubConduitShard.ts (hunk 1 above)
Comment (serycodes) [reply packages/api/src/endpoints/eventSub/HelixEventSubConduitShard.ts:44]: I copied them over from HelixEventSubSubscription.ts. Nothing currently is using them for conduits iirc.

Comment (d-fischer) [reply packages/api/src/endpoints/eventSub/HelixEventSubConduitShard.ts:44]: No need to have private properties that are not used for anything then.

File: packages/api/src/interfaces/endpoints/eventSub.external.ts (hunk 2 above)
Comment (d-fischer) [reply packages/api/src/interfaces/endpoints/eventSub.external.ts:104]: It just seems that the only place this type is used is in the conduit shards response, so having the conduit transport type here doesn't seem to make sense.

Also, your example uses private properties from the outside. Please never write production code like this, as there are no guarantees about these, ever.

Comment (serycodes) [reply packages/api/src/interfaces/endpoints/eventSub.external.ts:104]: It's on [HelixEventSubSubscriptionData](https://github.com/twurple/twurple/blob/2b02bf274c70efb968952d0597cfd89437591032/packages/api/src/interfaces/endpoints/eventSub.external.ts#L66) as well. This is an existing type already, I just added conduit data to it.

I see what you mean now though about nesting conduits. I'll fix that.

And yeah, that was just an example. I don't have anything in prod using private properties.

File: packages/api/src/index.ts [hunk 3]
```diff
@@ -85,18 +85,24 @@ export type {
 } from './interfaces/endpoints/entitlement.input';
 
 export { HelixEventSubApi } from './endpoints/eventSub/HelixEventSubApi';
+export { HelixEventSubConduit } from './endpoints/eventSub/HelixEventSubConduit';
+export { HelixEventSubConduitShard } from './endpoints/eventSub/HelixEventSubConduitShard';
 export { HelixEventSubSubscription } from './endpoints/eventSub/HelixEventSubSubscription';
 export { HelixPaginatedEventSubSubscriptionsRequest } from './endpoints/eventSub/HelixPaginatedEventSubSubscriptionsRequest';
 export type {
 	HelixEventSubTransportData,
+	HelixEventSubConduitShardData,
```
Comment (d-fischer) [inline packages/api/src/index.ts:94]: Please don't export external data interfaces here unless they are part of a publicly exposed parameter type, return type (both of these cases should be avoided though) or used in another Twurple package.

**Review by d-fischer — CHANGES_REQUESTED** [review]

**Review by d-fischer — APPROVED** [review]

**Pull request merged**